### PR TITLE
chore: use new build image

### PR DESCRIPTION
**Pull request by tomblench**
### Build changes to use new image

* update jenkinsfile to pipeline syntax

in order to build with java 17, quite a few changes needed to be done:

* upgrade gradlew to a recent version
* modernise build.gradle (especially the publishing steps)
* upgrade spotbugs
* fix a javadoc error
* update spring framework/boot versions (see below)
* update junit and mockito versions
* fix usage of now-removed `EnvironmentTestUtils` method

### About spring versions:

using Rich's suggestion:
```
I'm going to suggest Spring Framework 5.3.27 and Spring Boot 2.7.9
```
this way we can still support java 8.

### Testing

see green build https://cloudant-sdks-jenkins.swg-devops.com/job/Auxiliary/job/cloudant-spring/job/i192-use-new-build-image/13/

### CODE REVIEW

File: Jenkinsfile [hunk 1]
```diff
@@ -14,78 +14,90 @@
...
+        always {
+          sh "git config --unset credential.username || true"
+          sh "git config --unset credential.helper || true"
         }
       }
     }
-  }
-}
+  }  
+}
```
Comment (ricellis) [inline Jenkinsfile:103]: nit EOF

File: build.gradle [hunk 2]
```diff
@@ -13,187 +13,154 @@
...
-        throw new GradleException("Build failed due to javadoc warnings.");
-      }
+    group = 'com.cloudant'
+    version = new File(rootDir, 'VERSION').text.trim()
+
+    // If the version says "snapshot" anywhere assume it is not a release
+    ext.isReleaseVersion = !version.toUpperCase(Locale.ENGLISH).contains("SNAPSHOT")
+
+    sourceCompatibility = 11
+    targetCompatibility = 17
```
Comment (ricellis) [inline build.gradle:32]: Any particular reason to target 17 here? I agree that's fine from our test perspective, but in terms of binaries we ship I feel like it ought to be `11` (or even `8`)

**Review by ricellis — CHANGES_REQUESTED** [review]
Looks good, just a few nits

File: build.gradle (hunk 2 above)
Comment (tomblench) [reply build.gradle:32]: > As of Spring Framework 6.0, Spring requires Java 17+.

(from https://docs.spring.io/spring-framework/docs/current/reference/html/overview.html)

So I guess we just need to have a discussion about minimum spring/java versions we support.

Comment (ricellis) [reply build.gradle:32]: I'm going to suggest Spring Framework 5.3.27 and Spring Boot 2.7.9

Framework [5.3.x is documented](https://github.com/spring-projects/spring-framework/wiki/Spring-Framework-Versions#supported-versions) as:
>5.3.x is the final feature branch of the 5th generation, with long-term support provided on JDK 8, JDK 11, JDK 17 and the Java EE 8 level.

Boot [2.7.9 claims](https://docs.spring.io/spring-boot/docs/2.7.9/reference/htmlsingle/#getting-started.system-requirements):
>Spring Boot 2.7.9 requires Java 8 and is compatible up to and including Java 19. [Spring Framework 5.3.25](https://docs.spring.io/spring-framework/docs/5.3.25/reference/html/) or above is also required.

which I think should allow us to use Java 17, but with 1.8 source and target compatibility. I think that will give us slightly less churn and give us the best chance of utilizing the existing work from #22 and #23.

File: Jenkinsfile (hunk 1 above)
Comment (ricellis) [reply Jenkinsfile:99]: the trailing newline at the EOF is still missing

File: Jenkinsfile [hunk 3]
```diff
@@ -14,78 +14,90 @@
...
+      }
+    }
+    
+    // Publish the primary branch
+    stage('Publish') {
+      steps {
+        script {
+          if (env.BRANCH_IS_PRIMARY) {
+            checkout scm // re-checkout to be able to git tag
+            unstash name: 'built'
```
[the remark this answers is not in the document]
Comment (tomblench) [reply Jenkinsfile]: i think you're right, it shouldn't be necessary but the only way to know for sure is to do a release build

[the remark this answers is not in the document]
Comment (tomblench) [reply Jenkinsfile]: checkout line removed in 4aff060

**Review by ricellis — APPROVED** [review]
+1 but could you tweak https://github.com/cloudant-labs/cloudant-spring/blob/main/CONTRIBUTING.md to remove the info about gradle wrapper, thanks!

**Pull request merged**